plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/TransportReindexAction.java

```diff
@@ -0,0 +1,230 @@
+                super.copyRouting(index, doc);
+                break;
+            case "discard":
+                index.routing(null);
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported routing command");
+            }
+        }
+
+        /*
+         * Methods below here handle script updating the index request. They try
+         * to be pretty liberal with regards to types because script are often
+         * dynamically typed.
+         */
+        @Override
+        protected ReindexResponse buildResponse(long took) {
+            return new ReindexResponse(took, created(), updated(), batches(), versionConflicts(), noops(), failures());
+        }
+
+        @Override
+        protected void scriptChangedIndex(IndexRequest index, Object to) {
+            requireNonNull(to, "Can't reindex without a destination index!");
+            index.index(to.toString());
+        }
+
+        @Override
+        protected void scriptChangedType(IndexRequest index, Object to) {
+            requireNonNull(to, "Can't reindex without a destination type!");
+            index.type(to.toString());
+        }
+
+        @Override
+        protected void scriptChangedId(IndexRequest index, Object to) {
+            index.id(Objects.toString(to, null));
+        }
+
+        @Override
+        protected void scriptChangedVersion(IndexRequest index, Object to) {
+            if (to == null) {
+                index.version(Versions.MATCH_ANY).versionType(INTERNAL);
+                return;
+            }
+            index.version(asLong(to, "_version"));
+        }
+
+        @Override
+        protected void scriptChangedParent(IndexRequest index, Object to) {
+            // Have to override routing with parent just in case its changed
+            String routing = Objects.toString(to, null);
+            index.parent(routing).routing(routing);
+        }
+
+        @Override
+        protected void scriptChangedRouting(IndexRequest index, Object to) {
+            index.routing(Objects.toString(to, null));
+        }
+
+        @Override
+        protected void scriptChangedTimestamp(IndexRequest index, Object to) {
+            index.timestamp(Objects.toString(to, null));
+        }
+
+        @Override
+        protected void scriptChangedTtl(IndexRequest index, Object to) {
+            if (to == null) {
+                index.ttl(null);
+                return;
+            }
+            index.ttl(asLong(to, "_ttl"));
+        }
+
+        private long asLong(Object from, String name) {
+            try {
+                /*
+                 * Stuffing a number into the map will have converted it to
+                 * some Number.
+                 */
+                Number fromNumber = (Number) from;
+                long l = fromNumber.longValue();
+                // Check that we didn't round when we fetched the value.
+                if (fromNumber.doubleValue() != l) {
```

Comment:
Thanks for the reasoning. I just felt uneasy about comparing `#doubleValue()'  with `#longValue()'.